torch_geometric/nn/models/label_usage.py

```diff
@@ -0,0 +1,108 @@
+        split_ratio: float = 0.5,
+        num_recycling_iterations: int = 0,
+        return_tuple: bool = False,
+    ):
+
+        super().__init__()
+        self.base_model = base_model
+        self.num_classes = num_classes
+        self.split_ratio = split_ratio
+        self.num_recycling_iterations = num_recycling_iterations
+        self.return_tuple = return_tuple
+
+    def forward(
+        self,
+        feat: Tensor,
+        edge_index: Adj,
+        y: Tensor,
+        mask: Tensor,
+        val_idx: Tensor,
+        test_idx: Tensor,
+    ):
+        r"""Forward pass using label usage algorithm.
+
+        Args:
+            feat (torch.Tensor): Node feature tensor of dimension (N,F)
+                where N is the number of nodes and
+                F is the number of features per node
+            edge_index (torch.Tensor or SparseTensor): The edge indices with
+                dimension (N,N) for a dense adjacency matrix or (2,E) for a
+                sparse adjacency representation where E is the number of edges
+            y (torch.Tensor): Node ground-truth labels tensor of dimension
+                of (N,) for 1D tensor or (N,1) for 2D tensor
+            mask (torch.Tensor): Global indices of all nodes in training
+                set or mini-batch with dimension (N_train,) where N_train
+                is the number of nodes in training set or mini-batch
+            val_idx (torch.Tensor): Node indices in validation set with
+                dimension (N_val,) where N_val is the number of nodes
+                in validation set
+            test_idx (torch.Tensor): Node indices in test set with dimension
+                (N_test,) where N_test is the number of nodes in test set
+        """
+        assert feat.dim() == 2, "feat must be 2D but got shape {feat.shape}"
+
+        # random split mask based on split ratio
+        split_mask = torch.rand(mask.shape) < self.split_ratio
+        train_labels_idx = mask[split_mask]  # D_L: nodes with labels
+        train_pred_idx = mask[~split_mask]  # D_U: nodes to predict labels
+
+        assert y.dim() == 1 or (y.dim() == 2 and y.size(1) == 1),\
+            "Expected y to be either (N,) or (N, 1), but got shape {y.shape}"
+        # add labels to features for train_labels_idx nodes
+        # zero value nodes in train_pred_idx
+        onehot = torch.zeros([feat.shape[0], self.num_classes]).to(feat.device)
+        # create a one-hot encoding according to tensor dim
+        if y.dim() == 2:
+            onehot[train_labels_idx, y[train_labels_idx, 0]] = 1
+        else:
+            onehot[train_labels_idx, y[train_labels_idx]] = 1
+        feat = torch.cat([feat, onehot], dim=-1)
+
+        # set predictions for all unlabeled indices
+        unlabeled_idx = torch.cat([train_pred_idx, val_idx, test_idx])
+        # label reuse procedure
+        pred = self.base_model(feat, edge_index)
+        for _ in range(max(1, self.num_recycling_iterations + 1)):
```

Comment:
Now that you've moved the first iteration out of the for loop this should just be "range(self.num_recycling_iterations)`.  As is this will do one iteration of recycling even when user specifies `num_recylcing_iterations=0`